pyiron_contrib/generic/s3io.py

```diff
@@ -0,0 +1,567 @@
+                        if dict: dictionary containing the login credentials.
+                        if boto3.resource: This resource is used as is, requires bucket_name to be specified.
+                        if None: use Credentials as obtained by boto3, requires bucket_name to be specified.
+                bucket_name(str/None): Name of the bucket, overwrites name given in the config.
+
+            The configuration needs to provide the following information:
+                access_key or aws_access_key_id (str)
+                secret_key or aws_secret_access_key (str)
+            And may contain these additional fields:
+                endpoint or endpoint_url (str)
+                bucket (str)
+            Each additional keyword is passed on to boto3 as is.
+        """
+        self.bucket_name = None
+        self.s3resource = self._init_s3resource(credentials)
+
+        self.bucket_name = bucket_name or self.bucket_name
+        self._check_for_bucket_name()
+        self.bucket = self.s3resource.Bucket(self.bucket_name)
+
+    def _init_s3resource(self, credentials):
+        if isinstance(credentials, boto3.resources.base.ServiceResource):
+            return credentials
+        elif credentials is None:
+            # boto3 looks for the (missing) credentials at  ~/.aws/credentials or at environment variables:
+            # AWS_ACCESS_KEY_ID  AWS_SECRET_ACCESS_KEY  etc.
+            return boto3.resource('s3')
+        else:
+            credentials = self._get_credentials_dict(credentials)
+
+            config = credentials.pop('config', Config(s3={'addressing_style': 'path'}))
+
+            self._normalize_key(credentials, 'aws_access_key_id', 'access_key')
+            self._normalize_key(credentials, 'aws_secret_access_key', 'secret_key')
+            self._normalize_key(credentials, 'endpoint_url', 'endpoint')
+
+            self.bucket_name = credentials.pop("bucket", None)
+
+            return boto3.resource('s3', config=config, **credentials)
+
+    def _check_for_bucket_name(self):
+        if self.bucket_name is None:
+            raise ValueError("Bucket name needs to be provided.")
+
+    @staticmethod
+    def _normalize_key(credentials, key, alias_key):
+        value = credentials.pop(alias_key, None) or credentials.pop(key, None)
+        if value is not None:
+            credentials[key] = value
+
+    @staticmethod
+    def _get_credentials_dict(credentials):
+        if isinstance(credentials, str):
+            with open(credentials) as json_file:
+                credentials = json.load(json_file)
+        if not isinstance(credentials, dict):
+            raise TypeError(f"credentials is not one of the supported types but {type(credentials)}.")
+        return credentials.copy()
+
+    @property
+    def endpoint_url(self):
+        return self.s3resource.meta.client.meta.endpoint_url
+
+
+class FileS3IO:
```

Comment:
Some of the docstrings are indented one level too much.

Short class docstring with the main methods `download`/`upload`/`put`/... would be nice.